### DISCUSSION TITLE
Best way to display HTML emails?

### DISCUSSION THREAD

**Discussion by MetacodeX7** [AvaloniaUI/Avalonia#10428] [Q&A] [answered]
Hi,

I am new to Avalonia, and could use some advice on the best way to proceed. I am trying to build a mobile email client, but I've been encountering a lot of obstacles when it comes to displaying html emails.

- HtmlRenderer apparently doesn't support CSS, which I need. Didn't manage to make it work anyway.
- Outsystems.WebView worked in a desktop solution with Avalonia 0.10.18; but it doesn't support Avalonia 11 preview yet, and the cross platform solution requires that version (I think). There's also a limitation on processor type.
- Recently I also came across this post suggesting Microsoft's WebView2 could be directly used in Avalonia, but I haven't found any good example or explanation of NativeControlHost or why it's needed here.

I've explored a lot of directions and not even sure what is the right way to proceed in Avalonia. What would you say is the best solution to the problem? Is there another way to display Html content with CSS, without a WebView? Is there a plan to have an official cross-platform Webview in Avalonia?

Thanks,
Michael

**Comment by maxkatz6** [comment, marked as answer]
Hi!

> Is there another way to display Html content with CSS, without a WebView?

Not realistically, as it's a pretty hard job to copy HTML+CSS rendering engine.
Technically, HtmlRenderer could be able to render simple emails as per https://mailchimp.com/help/limitations-of-html-email/
But not everything will be supported anyway, as even original HtmlRenderer wasn't maintained/improved since ~2014 (I want to upgrade Avalonia fork, but original code still lacking).

> Recently I also came across this post suggesting Microsoft's WebView2 could be directly used in Avalonia, but I haven't found any good example or explanation of NativeControlHost or why it's needed here.


It can. But if you need a mobile (which I assumed from "mobile email client" part), it won't help. WebView2 as for now is Windows only. I think they had plans to ship macOS package too.

For the WebView2 (and WkWebView for macOS) usage in the Avalonia you can check https://github.com/maxkatz6/AvaloniaWebView
And for mobile, it can be done in the same way, except you need to use native controls specific to the platform. For example, in case of Android, we had this sample: https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog.Android/EmbedSample.Android.cs#L17-L20 and iOS https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog.iOS/EmbedSample.iOS.cs#L18-L21 (similar to macOS)

**Reply by maxkatz6** [reply]
But keep in mind, unlikely HtmlRenderer, any native control embedding in the avalonia will have an "airspace problem". The same as in WPF https://www.essentialobjects.com/forum/postst12065_WebBrowser-always-displaying-over-all-other-WPF-controls.aspx. I.e., no avalonia content can be rendered on top of native control.

**Reply by MetacodeX7** [reply]
Thank you for the detailed answer! I will try and reproduce the Android sample, and see if I can work with it despite the airspace issue.
Is WPF the only way to embed a native control? Or is there any way to fix the airspace issue if it does become a problem for my app?

**Reply by maxkatz6** [reply]
> Is WPF the only way to embed a native control

What do you mean? You don't need WPF to embed a native control. Although, you can embed WPF window into avalonia.

> Or is there any way to fix the airspace issue if it does become a problem for my app?

Not really.

**Reply by MetacodeX7** [reply]
Ah sorry, I misunderstood the part about WPF, I thought you meant native control embeddings were done using WPF. All those UI tech acronyms are new to me and they're all getting mixed up in my brain right now :)
